### ISSUE TITLE
Difference from Linux in file access checks

### ISSUE THREAD

**Issue by mirabilos** [google/gvisor#152]
Same environment as in #149, although I’ve created a custom image for this:

```
# cat >Dockerfile
FROM debian:buster
RUN apt-get update && \
    DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y tini mksh perl less ed screen
# docker build -t ti .
Sending build context to Docker daemon  770.6kB
Step 1/2 : FROM debian:buster
 ---> 2a853bf33864
Step 2/2 : RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y tini mksh perl less ed screen
 ---> Running in 6a2966773966
[…]
Processing triggers for libc-bin (2.28-7) ...
Removing intermediate container 6a2966773966
 ---> 6db34083702c
Successfully built 6db34083702c
Successfully tagged ti:latest
# docker run --runtime=runsc -it ti /usr/bin/tini /bin/bash
root@d916fb40ad74:/# cd /usr/share/doc/mksh
root@d916fb40ad74:/usr/share/doc/mksh# gzip -d check.*
root@d916fb40ad74:/usr/share/doc/mksh# sed -i s/@utflocale@/C.UTF-8/ check.t
root@d916fb40ad74:/usr/share/doc/mksh# perl check.pl -s check.t -p /bin/mksh -v -C shell:legacy-no,int:32,shell:textmode-no,shell:binmode-yes,stdout-ed,no-stderr-ed,regress:no-ctty
[…]
pass check.t:xxx-status-2
FAIL check.t:xxx-stat-1
        Description:
                Check that tests on files are consistent
        unexpected stdout - first difference: line 3, char 4 (wanted 0x31 ('1'), got 0x30 ('0')
        wanted: 
                1e 0 .
                1f 0 .
                2e 1 .
                2f 1 .
                3e 0 .
                3f 0 .
        got:
                1e 0 .
                1f 0 .
                2e 0 .
                2f 0 .
                3e 0 .
                3f 0 .
pass check.t:xxx-clean-chars-1
[…]
pass check.t:fsh-flags
Total failed: 1 (1 unexpected)
Total passed: 523
```

(The `regress:no-ctty` is necessary due to #151 as all the tests needing a controlling tty will fail. (`mknod /dev/tty c 5 0` will also fail.))

## tl;dr

The test fails because of this:

        mkdir a
        echo x >a/b
        chmod 0 a
        test -e a/b; echo 2e $? .
        test -f a/b; echo 2f $? .

Both `test -e` _and_ `test -f` are expected to fail on `a/b` because the user has no permissions to access the directory `a`.

**Comment by mirabilos** [comment]
… oh, wait… I just realised that this runs as root, and I’ve rechecked outside of the chroot, and the test will fail there as well. Oops, closing.

This means that the mksh testsuite for a binary built *outside* docker/gvisor will work, if you get it to working at all (see #149, #150) and only the tests that don’t need a controlling tty (see #151). This is a pretty good coverage of Linux syscalls.

For the record, gvisor can be detected by uname:

    Linux XXX.lan.tarent.de 4.19.0-3-amd64 #1 SMP Debian 4.19.20-1 (2019-02-11) x86_64 GNU/Linux

vs.

    Linux 8fe76467037f 4.4 #1 SMP Sun Jan 10 15:06:54 PST 2016 x86_64 GNU/Linux

(The version is just “4.4”, and the date is in a weird american timezone.)

**Comment by amscanne** [comment]
Re: detection, yeah, we figure there are going to be many ways to detect gVisor. We don't provide anything official to do this, but don't assume it will be a secret to an application that really wants to know. (We could do things like randomize uname, but there's always going to be some other way to do it.)

Thanks for these reports! We will triage.

These are the tests found in the [mksh repository](https://github.com/MirBSD/mksh/blob/master/check.pl), correct?

**Comment by mirabilos** [comment]
Adin Scannell dixit:

>Re: detection, yeah, we figure there are going to be many ways to
>detect gVisor. We don't provide anything official to do this, but
>don't assume it will be a secret to an application that really wants

OK, sure. It helps, anyway.

>Thanks for these reports! We will triage.
>
>These are the tests found in the [mksh
>repository](https://github.com/MirBSD/mksh/blob/master/check.pl),
>correct?

Indeed, they are.

They’ve proven a good compiler/toolchain/libc/kernel testsuite
as well, sometimes spotting OS bugs, often with signal delivery.
This was my entire idea, testing how good gvisor abstracts Linux.

But I closed this particular issue since it turns out it was
my fault. The only tests I could not successfully run out of
the entire testsuite were those that need a controlling tty.
So, good job so far!

bye,
//mirabilos
-- 
FWIW, I'm quite impressed with mksh interactively. I thought it was much
*much* more bare bones. But it turns out it beats the living hell out of
ksh93 in that respect. I'd even consider it for my daily use if I hadn't
wasted half my life on my zsh setup. :-) -- Frank Terbeck in #!/bin/mksh
